File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/WhoAmICommand.cs
using Atlasd.Daemon;
using Atlasd.Localization;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class WhoAmICommand : ChatCommand
    {
        public WhoAmICommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            var gameState = context.GameState;
            var ch = gameState.ActiveChannel;
            var g = gameState.GameAd;
            var r = ch == null ? Resources.YouAreUsingGameInRealm : Resources.YouAreUsingGameInTheChannel;

            if (g != null) // TODO: Consider private games and friendship.
                r = Resources.UserIsUsingGameInTheGame;
            else if (ch != null)
                r = Resources.UserIsUsingGameInTheChannel;
            else
                r = Resources.UserIsUsingGameInRealm;

            if (gameState.Away != null)
                r += Battlenet.Common.NewLine + Resources.AwayCommandStatusSelf.Replace("{awayMessage}", gameState.Away);

            var env = new Dictionary<string, string>()
            {
                { "accountName", gameState.Username },
                { "channel", ch == null ? "(null)" : ch.Name },
                { "game", Product.ProductName(gameState.Product, true) },
                { "gameAd", g == null ? "(null)" : Encoding.UTF8.GetString(g.Name) },
                { "host", Settings.GetString(new string[] { "battlenet", "realm", "host" }, "(null)") },
                { "localTime", gameState.LocalTime.ToString(Common.HumanDateTimeFormat).Replace(" 0", "  ") },
                { "name", Channel.RenderOnlineName(gameState, gameState) },
                { "onlineName", Channel.RenderOnlineName(gameState, gameState) },
                { "realm", Settings.GetString(new string[] { "battlenet", "realm", "name" }, Resources.Battlenet) },
                { "realmTime", DateTime.Now.ToString(Common.HumanDateTimeFormat).Replace(" 0", "  ") },
                { "realmTimezone", $"UTC{DateTime.Now:zzz}" },
                { "user", Channel.RenderOnlineName(gameState, gameState) },
                { "username", Channel.RenderOnlineName(gameState, gameState) },
                { "userName", Channel.RenderOnlineName(gameState, gameState) },
            };
            var en2v = env.Concat(context.Environment);

            foreach (var kv in env) r = r.Replace("{" + kv.Key + "}", kv.Value);
            new ChatEvent(ChatEvent.EventIds.EID_INFO, gameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, gameState.Ping, gameState.OnlineName, r).WriteTo(gameState.Client);
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/WhoAmICommand.vb
Imports AtlasV.Localization
Imports System
Imports System.Collections.Generic

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class WhoAmICommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            Dim ch = varContext.GameState.ActiveChannel
            Dim r = If(ch Is Nothing, Resources.YouAreUsingGameInRealm, Resources.YouAreUsingGameInTheChannel)

            If varContext.GameState.Away IsNot Nothing Then
                r += Battlenet.Common.NewLine + Resources.AwayCommandStatusSelf.Replace("{awayMessage}", varContext.GameState.Away)
            End If

            r = r.Replace("{channel}", If(ch Is Nothing, "(null)", ch.Name))
            r = r.Replace("{realm}", "BNETDocs")

            For Each kv In varContext.Environment
                r = r.Replace("{" & kv.Key & "}", kv.Value)
            Next

            Call New ChatEvent(ChatEvent.EventIds.EID_INFO, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, r).WriteTo(varContext.GameState.Client)
        End Sub
    End Class
End Namespace
